File: courses/focs-fa20/homeworks/homework4.ml
(* 

HOMEWORK 4

Due: Thu Oct 22, 2020 (23h59)

Name: 

Email:

Remarks, if any:

*)


(*
 *
 * Please fill in this file with your solutions and submit it
 *
 * The functions below are stubs that you should replace with your
 * own implementation.
 *
 * PLEASE DO NOT CHANGE THE TYPES IN THE STUBS BELOW.
 * Doing so risks making it impossible for me to test your code.
 *
 * Always make sure you can #use this file in a FRESH OCaml shell
 * before submitting it. It has to load without any errors.
 *
 *)



(*
 * Type for deterministic Turing machines
 *
 * States are integers
 *
 *)

type tm = {
    states : int list;
    input_alphabet : char list;
    tape_alphabet : char list;
    left_marker : char;
    blank : char;
    delta : (int * char) -> (int * char * int);   (* 0 = Left, 1 = Right *)
    start : int;
    accept : int;
    reject : int
  }

type config = {
    state : int;
    tape: char list;
    position: int
  }


(* 
 * Some sample deterministic Turing machines
 *
 * asbs is the regular language {a^m b^n | m,n >= 0}
 * anbn is the non-regular language {a^n b^n | n >= 0}
 * anbncn is the non-regular language {a^n b^n c^n | n >= 0}
 *
 *)

let asbs =
  let d inp = (match inp with
               | (1, 'a') -> (1, 'a', 1)
               | (1, 'b') -> (10, 'b', 1)
               | (1, '>') -> (1, '>', 1)
               | (1, '_') -> (777, '_', 1)
               | (10, 'b') -> (10, 'b', 1)
               | (10, '_') -> (777, '_', 1)
               | (777, 'a') -> (777, 'a', 1)
               | (777, 'b') -> (777, 'b', 1)
               | (777, '>') -> (777, '>', 1)
               | (777, '_') -> (777, '_', 1)
               | (_,c) -> (666,c,1))
  in { states = [1; 10; 777; 666];
       input_alphabet = ['a';'b'];
       tape_alphabet = ['a';'b';'_';'>'];
       blank = '_';
       left_marker = '>';
       start = 1;
       accept = 777;
       reject = 666;
       delta = d }

let anbn =
  let d inp = (match inp with
	       | (1, 'a') -> (1, 'a', 1)
     	       | (1, 'b') -> (10, 'b', 1)
	       | (1, '|') -> (1, '|', 1)
	       | (1, '-') -> (11, '-', 1)
	       | (10, 'b') -> (10, 'b', 1)
	       | (10, '.') -> (11, '.', 1)
	       | (11, '|') -> (12, '|', 1)
	       | (11, 'a') -> (11, 'a', 0)
	       | (11, 'b') -> (11, 'b', 0)
	       | (11, 'X') -> (11, 'X', 0)
	       | (11, '.') -> (11, '.', 0)
	       | (12, 'X') -> (12, 'X', 1)
	       | (12, '.') -> (777, '.', 1)
	       | (12, 'a') -> (13, 'X', 1)
	       | (13, 'a') -> (13, 'a', 1)
	       | (13, 'X') -> (13, 'X', 1)
	       | (13, 'b') -> (11, 'X', 1)
	       | (777, 'a') -> (777, 'a', 1)
	       | (777, 'b') -> (777, 'b', 1)
	       | (777, '|') -> (777, '|', 1)
	       | (777, 'X') -> (777, 'X', 1)
	       | (777, '.') -> (777, '.', 1)
	       | (_,c) -> (666,c,1))
  in { states = [1; 10; 11; 12; 13; 777; 666];
       input_alphabet = ['a';'b'];
       tape_alphabet = ['a';'b';'X';'.';'|'];
       blank = '.';
       left_marker = '|';
       start = 1;
       accept = 777;
       reject = 666;
       delta = d }
   

let anbncn =
  let d inp = (match inp with
	       | (1, 'a') -> (1, 'a', 1)
     	       | (1, 'b') -> (10, 'b', 1)
	       | (1, 'c') -> (15, 'c', 1)
	       | (1, '>') -> (1, '>', 1)
	       | (1, '_') -> (11, '_', 1)
	       | (10, 'b') -> (10, 'b', 1)
	       | (10, 'c') -> (15, 'c', 1)
	       | (10, '_') -> (11, '_', 1)
	       | (11, '>') -> (12, '>', 1)
	       | (11, 'a') -> (11, 'a', 0)
	       | (11, 'b') -> (11, 'b', 0)
	       | (11, 'c') -> (11, 'c', 0)
	       | (11, '_') -> (11, '_', 0)
	       | (11, 'X') -> (11, 'X', 0)
	       | (12, 'X') -> (12, 'X', 1)
	       | (12, '_') -> (777, '_', 1)
	       | (12, 'a') -> (13, 'X', 1)
	       | (13, 'a') -> (13, 'a', 1)
	       | (13, 'X') -> (13, 'X', 1)
	       | (13, 'b') -> (14, 'X', 1)
	       | (14, 'b') -> (14, 'b', 1)
	       | (14, 'X') -> (14, 'X', 1)
	       | (14, 'c') -> (11, 'X', 1)
	       | (15, 'c') -> (15, 'c', 1)
	       | (15, '_') -> (11, '_', 1)
	       | (777, 'a') -> (777, 'a', 1)
	       | (777, 'b') -> (777, 'b', 1)
	       | (777, 'c') -> (777, 'c', 1)
	       | (777, '>') -> (777, '>', 1)
	       | (777, 'X') -> (777, 'X', 1)
	       | (777, '_') -> (777, '_', 1)
	       | (_,c) -> (666, c,1))
  in { states = [1;10;11;12;13;14;15;777;666];
       input_alphabet = ['a';'b';'c'];
       tape_alphabet = ['a';'b';'c';'X';'_';'>'];
       blank = '_';
       left_marker = '>';
       start = 1;
       accept = 777;
       reject = 666;
       delta = d }


(*
 * Helper functions
 *
 *   explode : string -> char list
 *      return the list of characters making up a string
 *
 *   printConfig: tm -> config -> 'a -> 'a
 *      print a configuration (including newline) to standard output
 *      and return a value
 *
 *   validateStates : tm -> bool
 *      check that all the states reachable from the start state
 *      appear in the set of states of the Turing machine
 *
 *)

let explode (str: string): char list = 
  let rec acc (index, result) = 
    if (index < 0) then
      result
    else
      acc (index - 1, (String.get str index) :: result)
  in
  acc (String.length (str) - 1, [])

let printConfig (m: tm) (c: config): unit =
  let maxStateLength a r = max (String.length (string_of_int a)) r in
  let mw = List.fold_right maxStateLength m.states 0 in
  let padding = max 0 (c.position + 1 - List.length c.tape) in
  let rec mkBlank k =
    match k with
    | 0 -> []
    | _ -> m.blank :: (mkBlank (k - 1)) in
  let tape' : char list = c.tape @ (mkBlank padding) in
  let str_state = string_of_int c.state in
  let _ = print_string (String.sub (str_state ^ (String.make mw ' ')) 0 mw) in
  let _ = print_string "  "  in
  let _ = List.iteri (fun i sym -> 
              if (i = c.position) then
                Printf.printf "[%c]" sym
	      else
                Printf.printf " %c " sym) tape'  in
  print_newline ()


let validateStates (m: tm):bool =
  let rec addTransitions q syms queue =
    match syms with
    | [] -> queue
    | sym::syms' -> let (p, _, _) = m.delta (q, sym) in
                    addTransitions q syms' (p :: queue) in
  let rec loop seen queue =
    match queue with
    | [] -> true
    | q::qs -> if not (List.mem q m.states) then
                 let str_q = string_of_int q in
                 failwith ("Reachable state " ^ str_q ^ " not defined in states list")
               else if not (List.mem q seen) then
                 let str_q = string_of_int q in
                 let _ = print_string ("Following state " ^ str_q ^ "\n") in
                 loop (q :: seen) (addTransitions q m.tape_alphabet qs)
               else
                 loop seen qs  in
  loop [] [m.start]
  



(* QUESTION 1 *)


let startConfig (m:tm) (w:string): config =
  failwith "Not implemented"


let acceptConfig (m: tm) (c: config): bool = 
  failwith "Not implemented"


let rejectConfig (m: tm) (c: config): bool = 
  failwith "Not implemented"


let rec replace_nth (lst: 'a list) (n: int) (s: 'a): 'a list =
  failwith "Not implemented"


let step (m: tm) (c: config): config =
  failwith "Not implemented"


let run (m: tm) (w: string): bool = 
  failwith "Not implemented"
  


(* QUESTION 2 *)


let dummyTM = { states = [0; 1];
		input_alphabet = ['x'];
		tape_alphabet = ['x'; '_'; '>'];
		blank = '_';
		left_marker = '>';
		start = 0;
		accept = 0;
		reject = 1;
		delta = (fun (x,y) -> (x, y, 0))}
            

let tm_q2_ab3 : tm = dummyTM


let tm_q2_not : tm = dummyTM


let tm_q2_and : tm = dummyTM


let tm_q2_plus1 : tm = dummyTM

                     
let tm_q2_duplicate : tm = dummyTM

File: courses/focs-fa24/homeworks/4/homework4.py
# --------------------------------------------------
#
# HOMEWORK 4
#
# Due: Tue, Oct 29 2024 (23h59)
#
# Name:
#
# Email:
#
# Remarks, if any:
#
#
#
# --------------------------------------------------
#
# Please fill in this file with your solutions and submit it
#
# The functions below are stubs that you should replace with your
# own implementation.
#
# PLEASE DO NOT CHANGE THE SIGNATURE IN THE STUBS BELOW.
# Doing so risks making it impossible for me to test your code.
#
# --------------------------------------------------


#
# Some sample context-free grammars.
#

GR_ANBN = {
    "nonterminals": ["S"],
    "alphabet": ["a", "b"],
    "rules": [
        ("S", ""),
        ("S", "aSb")
    ],
    "start": "S"
}

GR_ANBM = {
    "nonterminals": ["S", "T", "U"],
    "alphabet": ["a", "b"],
    "rules": [
        ("S", "TU"),
        ("T", ""),
        ("T", "aTb"),
        ("U", ""),
        ("U", "Ub")
    ],
    "start": "S"
}


# Try to apply rules to a given string.

def replace(str, pos, lhs, rhs):
    # Replace lhs with rhs at pos of string.
    return str[:pos] + rhs + str[pos + len(lhs):]

def apply_rule(str, pos, rs):
    # Try to apply a rule in rs to str at position pos
    result = []
    for (lhs, rhs) in rs:
        if str[pos:].startswith(lhs):
            res = replace(str, pos, lhs, rhs)
            result.append(res)
    return result

def apply_rules(rs, str):
    result = []
    for pos in range(len(str)):
        res = apply_rule(str, pos, rs)
        result.extend(res)
    return result


# Remove null rules to simplify generation.

def nullable(g, nonterm):
    for (lhs, rhs) in g['rules']:
        if lhs == nonterm:
            if not rhs:
                return True
            for sym in rhs:
                if sym in g['alphabet']:
                    continue
                if sym == lhs:
                    # Skip in case of recursion.
                    continue
                if not nullable(g, sym):
                    continue
            return True
    return False

def remove_nullable(g):
    large = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    nonterminals = [x for x in g['nonterminals']]
    def get_next():
        for t in large:
            if t not in nonterminals:
                nonterminals.append(t)
                return t
        raise Exception("Ran out of capital letters!")
    new_start = get_next()
    rules = g['rules']
    rules = [r for r in rules if r[1]]
    new_rules = []
    for (lhs, rhs) in rules:
        nullables = [i for (i, s) in enumerate(rhs) if nullable(g, s)]
        update = [rhs]
        for i in nullables:
            new_update = []
            for c in update:
                new_update.append(c)
                new_update.append(replace(c, i, "?", "."))
            update = new_update
        new_rules.extend([(lhs, rhs.replace(".", "")) for rhs in update])
    new_rules.append((new_start, g['start']))
    if nullable(g, g['start']):
        new_rules.append((new_start, ""))
    return {
        "alphabet": g['alphabet'],
        "nonterminals": nonterminals,
        "rules": new_rules,
        "start": new_start
    }
            
# Perform an iteratively deepening depth-first search of the rewrite tree.

def dfs_path(maxdepth, grammar, target):
    lt = len(target)
    q = [([grammar['start']], 0)]
    seen = set()
    while q:
        (path, d) = q[0]
        ###print(path, d)
        q = q[1:]
        if path:
            str = path[-1]
            if len(str) == lt and str == target:
                return path
            if str in seen:
                continue
            if d > maxdepth:
                seen.add(str)
                continue
            new_strs = apply_rules(grammar['rules'], str)
            ##print(new_strs)
            new_strs_d = [(path + [x], d + 1) for x in new_strs]
            ##print(new_strs_d)
            q = new_strs_d + q
            seen.add(str)
        else:
            raise Exception(f"Problem: empty path in dfs_path?")
    return []

def idfs_path(maxdepth, grammar, target, verbose=False):
    for d in range(1, maxdepth):
        if verbose:
            print(f"Searching - depth {d}")
        path = dfs_path(d, grammar, target)
        ##print(path)
        if path:
            return path

# Try to generate a string for a given grammar.

def check_cfg(g):
    # Check if a grammar is context free.
    for r in g['rules']:
        (lhs, rhs) = r
        if len(lhs) != 1:
            raise Exception(f"Rule {r} is not context free")
        for sym in rhs:
            if sym not in g['alphabet'] and sym not in g['nonterminals']:
                raise Exception(f"Symbol {sym} not defined")
    if g['start'] not in g['nonterminals']:
        raise Exception(f"Start symbol not a nonterminal")
    
def generate(grammar, str, maxdepth):
    check_cfg(grammar)
    g = remove_nullable(grammar)
    path = idfs_path(maxdepth, g, str, verbose=True)
    for (i, p) in enumerate(path or []):
        print(f"{'->' if i else ''} {p}")
    return bool(path)


#
# QUESTION 1
#

GR_A = None

GR_B = None

GR_C = None

GR_D = None




# Sample Turing machines.


TM_ANBN = {
    "states": [1, 2, 4, 6, 7, 777, 666],
    "alphabet": ['a','b'],
    "tape_alphabet": ['a','b','X','Y','_'],
    "start": 1,
    "accept": 777,
    "reject": 666,
    "delta": [
        (1, 'a', 2, 'X', 1),
        (1, '_', 777, '_', 1),
        (2, 'a', 2, 'a', 1),
        (2, 'Y', 2, 'Y', 1),
        (2, 'b', 4, 'Y', -1),
        (4, 'Y', 4, 'Y', -1),
        (4, 'a', 7, 'a', -1),
        (4, 'X', 6, 'X', 1),
        (6, 'Y', 6, 'Y', 1),
        (6, '_', 777, '_', 1),
        (7, 'a', 7, 'a', -1),
        (7, 'X', 1, 'X', 1)
    ]
}


TM_ANBNCN = {
    "states": [1, 2, 3, 4, 5, 6, 7, 8, 666, 777],
    "alphabet": ['a','b','c'],
    "tape_alphabet": ['a','b','c','X','Y','Z','_'],
    "start": 1,
    "accept": 777,
    "reject": 666,
    "delta": [
        (1, 'a', 2, 'X', 1),
        (1, '_', 777, '_', 1),
        (2, 'a', 2, 'a', 1),
        (2, 'Y', 2, 'Y', 1),
        (2, 'b', 3, 'Y', 1),
        (3, 'b', 3, 'b', 1),
        (3, 'Z', 3, 'Z', 1),
        (3, 'c', 4, 'Z', -1),
        (4, 'Z', 4, 'Z', -1),
        (4, 'Y', 5, 'Y', -1),
        (4, 'b', 7, 'b', -1),
        (5, 'Y', 5, 'Y', -1),
        (5, 'X', 6, 'X', 1),
        (6, 'Y', 6, 'Y', 1),
        (6, 'Z', 6, 'Z', 1),
        (6, '_', 777, '_', 1),
        (7, 'b', 7, 'b', -1),
        (7, 'Y', 7, 'Y', -1),
        (7, 'a', 8, 'a', -1),
        (8, 'a', 8, 'a', -1),
        (8, 'X', 1, 'X', 1)
    ]
}


# Print a configuration from a Turing machine.

def print_config(m, c):
    (state, tape, pos) = c
    width = max(len(str(q)) for q in m['states'])
    padding = max(0, pos - len(tape))
    tape = tape + (' ' * padding)
    content = "".join(f"[{x}]" if i + 1 == pos else f" {x} " for (i, x) in enumerate(tape))
    print(f"{(str(state) + ' ' * width)[:width]} {content}")


#
# QUESTION 2
#

def start_tm(m, w):
    raise Exception("Not implemented")


def is_accept_tm(m, c):
    raise Exception("Not implemented")

    
def is_done_tm(m, c):
    raise Exception("Not implemented")

    
def step_tm(m, c):
    raise Exception("Not implemented")


def accept_tm(m, input):
    raise Exception("Not implemented")



#
# QUESTION 3
#

TM_EQUAL = None

TM_AND = None

TM_PLUS1 = None

TM_PLUS = None

